cpp/src/parquet/encryption/properties_driven_crypto_factory.h

```diff
@@ -0,0 +1,210 @@
+
+   private:
+    std::string footer_key_;
+    std::string column_keys_;
+    ParquetCipher::type encryption_algorithm_;
+    bool plaintext_footer_;
+    bool double_wrapping_;
+    uint64_t cache_lifetime_seconds_;
+    bool internal_key_material_;
+    bool uniform_encryption_;
+    int32_t data_key_length_bits_;
+  };
+
+  const std::string& footer_key() const { return footer_key_; }
+  const std::string& column_keys() const { return column_keys_; }
+  ParquetCipher::type encryption_algorithm() const { return encryption_algorithm_; }
+  bool plaintext_footer() const { return plaintext_footer_; }
+  bool double_wrapping() const { return double_wrapping_; }
+  uint64_t cache_lifetime_seconds() const { return cache_lifetime_seconds_; }
+  bool internal_key_material() const { return internal_key_material_; }
+  bool uniform_encryption() const { return uniform_encryption_; }
+  int32_t data_key_length_bits() const { return data_key_length_bits_; }
+
+ private:
+  EncryptionConfiguration(const std::string& footer_key, const std::string& column_keys,
+                          ParquetCipher::type encryption_algorithm, bool plaintext_footer,
+                          bool double_wrapping, uint64_t cache_lifetime_seconds,
+                          bool internal_key_material, bool uniform_encryption,
+                          int32_t data_key_length_bits)
+      : footer_key_(footer_key),
+        column_keys_(column_keys),
+        encryption_algorithm_(encryption_algorithm),
+        plaintext_footer_(plaintext_footer),
+        double_wrapping_(double_wrapping),
+        cache_lifetime_seconds_(cache_lifetime_seconds),
+        internal_key_material_(internal_key_material),
+        uniform_encryption_(uniform_encryption),
+        data_key_length_bits_(data_key_length_bits) {}
+
+  std::string footer_key_;
+  std::string column_keys_;
+  ParquetCipher::type encryption_algorithm_;
+  bool plaintext_footer_;
+  bool double_wrapping_;
+  uint64_t cache_lifetime_seconds_;
+  bool internal_key_material_;
+  bool uniform_encryption_;
+  int32_t data_key_length_bits_;
+};
+
+class PARQUET_EXPORT DecryptionConfiguration {
+ public:
+  class PARQUET_EXPORT Builder {
+   public:
+    Builder() : cache_lifetime_seconds_(kDefaultCacheLifetimeSeconds) {}
+
+    /// Lifetime of cached entities (key encryption keys, local wrapping keys, KMS client
+    /// objects).
+    /// The default is 600 (10 minutes).
+    Builder* cache_lifetime_seconds(uint64_t cache_lifetime_seconds);
+
+    std::shared_ptr<DecryptionConfiguration> build();
+
+   private:
+    uint64_t cache_lifetime_seconds_;
+  };
+
+  uint64_t cache_lifetime_seconds() const { return cache_lifetime_seconds_; }
+
+ private:
+  explicit DecryptionConfiguration(uint64_t cache_lifetime_seconds)
+      : cache_lifetime_seconds_(cache_lifetime_seconds) {}
+
+  uint64_t cache_lifetime_seconds_;
+};
+
+class PARQUET_EXPORT PropertiesDrivenCryptoFactory {
```

Comment:
The name is based on Java version: https://github.com/apache/parquet-mr/pull/615/files#diff-b0ae7da2b29e6a52fd3b3a91faeb02e1 and "properties driven", as I know, links with hadoop properties: https://github.com/apache/parquet-mr/pull/615/files#diff-cb88582794e86857f46fa3a3aff7ec40R36. @ggershinsky and I think we should keep the same API as much as possible between Java and C++.

Reply:
As said above, I don't think keeping the API the same has much value. Different languages have different favored conventions. There is much to be said about Java's love of long names...

Reply:
What about `PropertiesCryptoFactory`?

Reply:
Perhaps. I can't really give an opinion since I don't really understand what the purpose of this class it, sorry. I may be mistaken, but it seems to be a mix of unrelated functionality.

Reply:
sorry about that. This is a core class, that translates the parameters of high level encryption (like the names of encrypted columns, names of master keys, etc), into parameters of low level encryption (like the key metadata, DEK, etc). A factory that produces the low level FileEncryptionProperties and FileDecryptionProperties objects, from the high level parameters. I agree though the name is long and confusing. Moreover, we don't have another crypto factory here, so something simple like `CryptoFactory` would do as the class name from my POV.

Reply:
`CryptoFactory` sounds good. You can also add a simple docstring explaining what you just said :-)

Reply:
We'll do :)